proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,355 @@
+```kotlin
+class C {
+    val elementList: List<Element>
+        field = mutableListOf()
+}
+```
+
+## Table of contents
+
+<!--- TOC -->
+
+* [Use Cases](#use-cases)
+  * [Expose read-only subtype](#expose-read-only-subtype)
+  * [Decouple storage type from external representation](#decouple-storage-type-from-external-representation)
+  * [Expose read-only view](#expose-read-only-view)
+  * [Access field from outside of getter and setter](#access-field-from-outside-of-getter-and-setter)
+* [Design](#design)
+  * [Explicit Backing Fields](#explicit-backing-fields)
+    * [Restrictions](#restrictions)
+    * [Accessors](#accessors)
+    * [Visibility](#visibility)
+    * [Lateinit](#lateinit)
+  * [Smart Type Narrowing](#smart-type-narrowing)
+* [Alternatives](#alternatives)
+  * [Initial Proposal](#initial-proposal)
+* [Future Enhancements](#future-enhancements)
+  * [Direct Backing Field Access](#direct-backing-field-access)
+  * [Protected Fields](#protected-fields)
+  * [Mutable Fields for Read-only Properties](#mutable-fields-for-read-only-properties)
+
+<!--- END -->
+
+## Use Cases
+
+This proposal caters to a variety of use-cases that are currently met via a backing property pattern.
+
+### Expose read-only subtype
+
+We often do not want our data structures to be modified from outside. It is customary in Kotlin to have 
+a read-only (e.g. `List`) and a mutable (e.g. `MutableList`) interface to the same data structure. 
+
+```kotlin
+internal val _items = mutableListOf<Item>()
+val item : List<Item> by _items
+```
+
+And the new syntax allows us to write:
+
+```kotlin
+val items: List<Item>
+    internal field = mutableListOf()
+```
+
+This use-case is also widely applicable to architecture of reactive applications:
+ 
+* Android `LiveData` has a `MutableLiveData` counterpart.
+* Rx `Observable` has a mutable `Subject` counterpart.
+* Kotlin coroutines `SharedFlow` has a `MutableSharedFlow`, etc.
+
+For example, sample code [from an Android app](https://github.com/elpassion/crweather/blob/9c3e3cb803b7e4fffbb010ff085ac56645c9774d/app/src/main/java/com/elpassion/crweather/MainModel.kt#L14):
+
+```kotlin
+private val mutableCity = MutableLiveData<String>().apply { value = "" }
+private val mutableCharts = MutableLiveData<List<Chart>>().apply { value = emptyList() }
+private val mutableLoading = MutableLiveData<Boolean>().apply { value = false }
+private val mutableMessage = MutableLiveData<String>()
+
+val city: LiveData<String> = mutableCity
+val charts: LiveData<List<Chart>> = mutableCharts
+val loading: LiveData<Boolean> = mutableLoading
+val message: LiveData<String> = mutableMessage
+```
+
+Becomes:
+
+```kotlin
+val city: LiveData<String>
+    field = MutableLiveData().apply { value = "" }
+val charts: LiveData<List<Chart>>
+    field = MutableLiveData().apply { value = emptyList() }
+val loading: LiveData<Boolean>
+    field = MutableLiveData().apply { value = false }
+val message: LiveData<String>
+    field = MutableLiveData()
+```
```

Comment:
It can be noted that while the total number of lines stays roughly the same, the related properties and fields are better grouped together, and also the repetition in property names is avoided.

Reply:
Added a note to that effect.